mempool/src/shared_mempool.rs

```diff
@@ -259,6 +269,158 @@ async fn process_incoming_transactions<V>(
     notify_subscribers(SharedMempoolNotification::NewTransactions, &smp.subscribers);
 }
 
+/// send callback here
+async fn process_rpc_submit_transactions_request<V>(
+    smp: SharedMempool<V>,
+    peer_id: PeerId,
+    request: SubmitTransactionsRequest,
+    callback: oneshot::Sender<Result<Bytes, RpcError>>,
+) where
+    V: TransactionValidation,
+{
+    // the RPC response object that will be returned to sender via callback
+    let mut response_msg = None;
+
+    if let Ok(response) = submit_transactions_to_mempool(smp, peer_id, request).await {
+        let smp_msg = MempoolSyncMsg {
+            message: Some(MempoolSyncMsg_oneof::SubmitTransactionsResponse(response)),
+        };
+        response_msg = Some(smp_msg);
+    }
+
+    if let Some(response_msg) = response_msg {
+        // send response to callback
+        let response_data = response_msg.to_bytes().expect("failed to serialize proto");
+        if let Err(err) = callback
+            .send(Ok(response_data))
+            .map_err(|_| format_err!("[shared mempool] handling inbound RPC call timed out"))
+        {
+            error!(
+                "[shared mempool] failed to process batched transaction request, error: {:?}",
+                err
+            );
+        }
+    } else {
+        error!(
+            "[shared mempool] Did not get a response msg back from submit transactions from upstream peer / peer validator"
+        );
+    }
+}
+
+/// Tries to add transactions to local mempool
+/// returns RPC response that can be sent to callback
+/// peer_id is the PeerId of the node that sent the request
+async fn submit_transactions_to_mempool<V>(
+    smp: SharedMempool<V>,
+    peer_id: PeerId,
+    request: SubmitTransactionsRequest,
+) -> failure::Result<SubmitTransactionsResponse>
+where
+    V: TransactionValidation,
+{
+    /////////////////////////////////////////////
+    // convert from proto to SignedTransaction //
+    /////////////////////////////////////////////
+    let transactions: Vec<_> = request
+        .transactions
+        .clone()
+        .into_iter()
+        .filter_map(|txn| match SignedTransaction::try_from(txn.clone()) {
+            Ok(t) => Some(t),
+            Err(e) => {
+                // TODO make RPC response for this invalid transaction
+                // log error
+                security_log(SecurityEvent::InvalidTransactionMP)
+                    .error(&e)
+                    .data(&txn)
+                    .log();
+                None
+            }
+        })
+        .collect();
+
+    //////////////////////////////////
+    // validate transactions via VM //
+    //////////////////////////////////
+
+    // eagerly filter out transactions that were already committed
```

Comment:
btw eventually we want to get rid of eager filtering. It should be unnecessary after either we move to use libra channel and keep only one message per peer or move it to new flow. I'd even remove it here